IPVS: kubelet, kube-proxy: unmark packets before masquerading …

aojea opened the pull request:
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation

/kind failing-test

> /kind feature
> /kind flake

**What this PR does / why we need it**:

The sig-network tests started to fail after the double nat fix.
The problem is that the IPVS proxier uses a POSTORUTING for the hairpin traffic that needs to happen before the
kubernetes specific postrouting rules. Prepending this ipvs only masquerade rule solves the issue.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #92553 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```


> @aojea catching up on this. So we're reverting the double-nat fix for IPVS only since it used ipset and was not impacted?

aojea commented:
/test pull-kubernetes-e2e-gci-gce-ipvs

aojea commented:
/assign @andrewsykim @uablrek 

This solves all the test failures reported in #92553 

```
I0628 16:24:49.389] FAIL! -- 757 Passed | 1 Failed | 0 Pending | 4355 Skipped
```

and gives us back to the original situation reported in, with the flexvolume test failure #91253

```
I0628 16:24:49.388] [Fail] [sig-storage] Flexvolumes [It] should be mountable when non-attachable 
I0628 16:24:49.388] test/e2e/framework/volume/fixtures.go:502
```

@danwinship added this fix to solve the double natting problem, I tried to backtrace the problem but seems that IPVS uses ipset too and I wasn't able to find out the problem , so reverting the change only in IPVS will solve the issue

aojea commented:
/priority urgent-critical
This breaks kube-proxy in IPVS mode

/test pull-kubernetes-e2e-kind-ipv6 
/test pull-kubernetes-e2e-gci-gce-ipvs

andrewsykim commented:
@aojea catching up on this. So we're reverting the double-nat fix for IPVS only since it used ipset and was not impacted?

aojea commented:
> @aojea catching up on this. So we're reverting the double-nat fix for IPVS only since it used ipset and was not impacted?

I don't know if it was impacted, sorry, didn't go so far understanding the ipset - iptables relationship, however, with the fix in  IPVS  is broken per the test results.

I think that we should follow up and verify if the double-nat fix is needed for IPVS, but have it working until then.

aojea commented:
/retest
it works :smile: 
> pull-kubernetes-e2e-gci-gce-ipvs — Job succeeded.

aojea commented:
/retest

Lion-Wei commented on test/e2e/storage/flexvolume.go:
```diff
@@ -182,7 +182,7 @@ var _ = utils.SIGDescribe("Flexvolumes", func() {
 	})
 
-	ginkgo.It("should be mountable when non-attachable", func() {
+	ginkgo.It("should be mountable when non-attachable [Feature:Flexvolumes]", func() {
```
I know this is for storage feature, but still feel inappropriate to directly add this label.

aojea replied:
This test is failing constantly, is breaking this job.

It seems a leftover from https://github.com/kubernetes/kubernetes/pull/91329, the linked issue says that these tests should have the feature tag to avoid breaking the e2e suite
https://github.com/kubernetes/kubernetes/issues/78737#issuecomment-627021732

I've added it as a separate commit so the git history is clean and can be reverted independently.

andrewsykim replied:
cc @msau42 is this okay to do here?

msau42 replied:
The switch the distroless should not have impacted the "non-attachable" test case. This test case is passing fine in the normal k8s job: https://k8s-testgrid.appspot.com/sig-storage-kubernetes#gce&include-filter-by-regex=Flexvolume

So it would be interesting to figure out why it is consistently failing in this job.

aojea replied:
sorry about that, my mistake

aojea commented:
/retest

Lion-Wei commented:
@aojea I think part of the problem is because `RETURN` rule in `KUBE-POSTROUTING` break the loop back logic. 
```
Chain KUBE-POSTROUTING (1 references)
target     prot opt source               destination
RETURN     all  --  0.0.0.0/0            0.0.0.0/0            mark match ! 0x4000/0x4000
MARK       all  --  0.0.0.0/0            0.0.0.0/0            MARK xor 0x4000
MASQUERADE  all  --  0.0.0.0/0            0.0.0.0/0            /* kubernetes service traffic requiring SNAT */
MASQUERADE  all  --  0.0.0.0/0            0.0.0.0/0            /* Kubernetes endpoints dst ip:port, source ip for solving hairpin purpose */ match-set KUBE-LOOP-BACK dst,dst,src
```

However, I think straightly revert the double-nat fix is not a good idea, maybe move `KUBE-LOOP-BACK` up could be an option.

aojea commented:
> However, I think straightly revert the double-nat fix is not a good idea, maybe move `KUBE-LOOP-BACK` up could be an option.


agree that the best solution is to solve it, but right now IPVS mode is broken and I don't know if the double-nat issue with the random-fully flag is affecting IPVS, but if it does affect, it's only in some specific scenarios, we can see that all the test succeed now.

Honestly, I don't have more time to investigate :sweat_smile: , seems this logic was added here https://github.com/kubernetes/kubernetes/pull/54219 , my suggestion is to merge this, fix the job and IPVS so we can have a good test signal, and open a new Issue to solve/understand the double-nat issue in IPVS.

Lion-Wei commented:
> > However, I think straightly revert the double-nat fix is not a good idea, maybe move `KUBE-LOOP-BACK` up could be an option.
> 
> agree that the best solution is to solve it, but right now IPVS mode is broken and I don't know if the double-nat issue with the random-fully flag is affecting IPVS, but if it does affect, it's only in some specific scenarios, we can see that all the test succeed now.
> 
> Honestly, I don't have more time to investigate 😅 , seems this logic was added here #54219 , my suggestion is to merge this, fix the job and IPVS so we can have a good test signal, and open a new Issue to solve/understand the double-nat issue in IPVS.

Agree.

danwinship commented:
There were definitely people using IPVS in various places in the sprawling tangle of bug reports related to the random-fully/checksum problem.

> I think part of the problem is because `RETURN` rule in `KUBE-POSTROUTING` break the loop back logic.

> maybe move KUBE-LOOP-BACK up could be an option.

er, yeah, my fix assumed that kubelet and kube-proxy were creating identical `KUBE-POSTROUTING` rules.

This will hopefully eventually get fixed as part of #82125 but yes, it seems like for now just moving the `KUBE-LOOP-BACK` rule should fix this

danwinship commented:
Also note that the unmark PR was backported back to 1.18, 1.17 and 1.16

aojea commented:
> Also note that the unmark PR was backported back to 1.18, 1.17 and 1.16

/hold
that changes everything then, we have to fix it so we can backport it

andrewsykim commented:
So I guess this also means IPVS is broken on the the v1.18, v1.17 and 1.16 release branches? 

> it seems like for now just moving the KUBE-LOOP-BACK rule should fix this

Do you think you can update this PR to do this?

andrewsykim commented:
cc @kubernetes/release-managers

aojea commented:
> So I guess this also means IPVS is broken on the the v1.18, v1.17 and 1.16 release branches?
> 
> > it seems like for now just moving the KUBE-LOOP-BACK rule should fix this
> 
> Do you think you can update this PR to do this?

@andrewsykim I can't understand the logic to move up the  KUBE-LOOP-BACK , I think that for someone more familiar with the code should be easy

andrewsykim commented:
I think we want to move some of the masq mark rules in `createAndLinkeKubeChain` to `writeIptablesRules`. Not sure yet if that's the right solution but that would move the ordering of the rules so KUBE-LOOP-BACK is first, let's see if the e2es pass with that change first.

andrewsykim commented:
@aojea actually it might be easier to update KUBE-LOOP-BACK to target KUBE-MARK-MASQ instead of MASQUERADE directly. This should be a one-line change here https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/ipvs/proxier.go#L171

andrewsykim commented:
^ nvm cause the early return from KUBE-POSTROUTING is still there, but we should still do this in a follow-up PR

rikatz commented:
Well my bad, I thought that I've tested this also in IPVS but it seems that not :/ Sorry about that

aojea commented:
> Well my bad, I thought that I've tested this also in IPVS but it seems that not :/ Sorry about that

this is nobodies fault

andrewsykim commented:
> this is nobodies fault

+100 

I'll add that this only impacts routing of pods back to the originating node so it is really easy to miss in testing. Good thing we had e2es to catch it :)

thockin commented:
Good catch.  @rikatz these are the sorts of things that just happen.  There are and will always be corner cases that are not as well-exposed as the main paths.  Don't sweat it :)

aojea committed:
kube-proxy ipvs masquerade hairpin traffic

Masquerade de traffic that loops back to the originator
before they hit the kubernetes-specific postrouting rules

Signed-off-by: Antonio Ojea <antonio.ojea.garcia@gmail.com>

aojea commented:
ok, let's see if instead of masquerading directly the traffic, we just detect it and mark it, avoiding the return rule

aojea commented:
this is how it look like now
```
-A PREROUTING -m comment --comment "kubernetes service portals" -j KUBE-SERVICES
-A OUTPUT -m comment --comment "kubernetes service portals" -j KUBE-SERVICES
-A POSTROUTING -m comment --comment "kubernetes postrouting rules" -j KUBE-POSTROUTING
-A KUBE-FIREWALL -j KUBE-MARK-DROP
-A KUBE-LOAD-BALANCER -j KUBE-MARK-MASQ
-A KUBE-MARK-MASQ -j MARK --set-xmark 0x4000/0x4000
-A KUBE-POSTROUTING -m mark ! --mark 0x4000/0x4000 -j RETURN
-A KUBE-POSTROUTING -j MARK --set-xmark 0x4000/0x0
-A KUBE-POSTROUTING -m comment --comment "kubernetes service traffic requiring SNAT" -j MASQUERADE --random-fully
-A KUBE-SERVICES -m comment --comment "Kubernetes endpoints dst ip:port, source ip for solving hairpin purpose" -m set --match-set KUBE-LOOP-BACK dst,dst,src -j KUBE-MARK-MASQ
-A KUBE-SERVICES ! -s 10.244.0.0/16 -m comment --comment "Kubernetes service cluster ip + port for masquerade purpose" -m set --match-set KUBE-CLUSTER-IP dst,dst -j KUBE-MARK-MASQ
-A KUBE-SERVICES -m addrtype --dst-type LOCAL -j KUBE-NODE-PORT
-A KUBE-SERVICES -m set --match-set KUBE-CLUSTER-IP dst,dst -j ACCEPT
```

aojea commented:
nope, is not working :/

aojea commented:
hehe, it works, I know that is not the more elegant fix but it works :smile: 
/hold cancel
/priority critical-urgent

aojea commented:
/retest

andrewsykim commented on pkg/proxy/ipvs/proxier.go:
```diff
@@ -1845,6 +1844,17 @@ func (proxier *Proxier) createAndLinkeKubeChain() {
 
+	// Handle traffic that loops back to the originator with SNAT.
+	// The rules should go before the kubernetes-specific ones
+	if !proxier.ipsetList[kubeLoopBackIPSet].isEmpty() {
```
I don't think this would work because this is before we actually insert entries into the ipset https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/ipvs/proxier.go#L1164-L1169

aojea replied:
the test are passing so it has to be working :sweat_smile: 
but yeah, it seems it can skip one cycle, I think it works because the ipset will have always the coredns endpoints so once it's written it's always going to be there

andrewsykim replied:
Yeah it may be "working" but I don't think this is the correct solution. On line 1081, we reset the ipset entries, but this happens after `createAndLinkeKubeChain` so here we are checking against the ipset from the previous sync.

aojea replied:
right, just was trying to understand why it worked, definitively we can not move ahead with this,.
I'll change it

andrewsykim replied:
yeah I'm curious too, I think we can assume it's from the previous sync but not 100% sure  :thinking:

andrewsykim commented:
@aojea could we instead move the mark masq rules added in https://github.com/kubernetes/kubernetes/pull/92035 to `writeIptableRules`?

aojea commented:
/retest

andrewsykim reviewed (commented):
/approve

Thanks @aojea :bow: 

@danwinship can you give this one another review to make sure the mark/unmark logic is correct. Logic is the same, just shifting the ordering so it's after any rules we add to KUBE-POSTROUTING in the IPVS proxy.

danwinship commented:
yup, lgtm

aojea commented:
/retest

andrewsykim commented:
/lgtm

andrewsykim commented:
@aojea are you gonna do the cherry-picks?

Lion-Wei commented:
Seems like problem perfectly solved. 😄
If we want cherry-pick this pr, maybe leave the flexvolume commit alone. 
/lgtm

msau42 commented:
/hold
sorry for the late response, but adding a feature tag to the flexvolume test is not ideal as it will not run in the default k8s jobs anymore.

aojea commented:
> /hold
> sorry for the late response, but adding a feature tag to the flexvolume test is not ideal as it will not run in the default k8s jobs anymore.

yeah, my fault, was assuming it was a leftover, let me remove it

aojea commented:
/hold cancel
flexvolume commit removed

aojea commented:
@Lion-Wei do you mind lgtm again?

aojea commented:
/retest
the ipvs job failure is because the flexvolume test only

Lion-Wei commented:
@aojea Not at all.
/lgtm

aojea commented:
/retest
:thinking:

andrewsykim commented:
I thought the flex volume tests were failing because quay.io was down but that seemed to be fixed now https://github.com/kubernetes/kubernetes/issues/91242 ?

andrewsykim commented:
Maybe we just need to update to use the staging CI image for quay? @BenTheElder might know?

andrewsykim reviewed (commented):
/hold cancel

Removing the hold though since only the flex volume tests are failing and this is a known issue

msau42 commented:
I don't think root cause of flex volume test failing in this job is known. It's passing in normal k8s ci, so I suspect it is related to the configuration of this job.

This line from the test log is potentially suspicious and indicates that the test is trying to use IPs.
```
Jul  1 15:50:26.384: INFO: Getting external IP address for e2e-bc0cae53a3-2e794-minion-group-mxnp
```
